File: Data Structures/Linked Lists/Circular Linked Lists/circular-linked-list.rb
# Circular Linked List

class Node
    attr_accessor :val, :next

    def initialize(val, next_node)
        @val = val
        @next = next_node
    end
end

class CircularLinkedList
    def initialize(val)
        @head = Node.new(val, nil)
    end

    # still to fix
    def push(new_data)
        ptr = Node.new(new_data, @head)
        temp = @head

        if @head != nil
            while(temp.next != @head)
                temp = temp.next
            end
            temp.next = ptr
        else
            ptr.next = ptr
        end
        @head = ptr
    end

    def print_list
        current = @head
        if @head != nil
            while(1)
                puts "#{current.val}"
                current = current.next
                if current == @head
                    break
                end
            end
        end
    end
end

l = CircularLinkedList.new(3)
l.push(4)
l.push(5)
l.push(6)
l.push(7)
l.print_list()

File: Data Structures/Linked Lists/Circular Linked Lists/circular-linked-list.py
"""
    Circular Linked List
"""


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class CircularLinkedList:
    def __init__(self):
        self.head = None

    def push(self, new_data):
        ptr = Node(new_data)
        temp = self.head
        ptr.next = self.head

        if self.head is not None:
            while(temp.next != self.head):
                temp = temp.next
            temp.next = ptr
        else:
            ptr.next = ptr
        self.head = ptr

    def printList(self):
        temp = self.head
        if self.head is not None:
            while(True):
                print("%d" % (temp.data))
                temp = temp.next
                if (temp == self.head):
                    break


if __name__ == '__main__':
    llist = CircularLinkedList()
    llist.push(7)
    llist.push(1)
    llist.push(8)
    print(llist.printList())
